drivers/clock_control/clock_agilex_ll.c

```diff
@@ -0,0 +1,392 @@
+/*
+ * Copyright (c) 2019-2021, Intel Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <arch/cpu.h>
+#include <drivers/clock_control/clock_agilex_ll.h>
+#include <errno.h>
+#include <soc.h>
+#include <socfpga_handoff.h>
+#include <socfpga_system_manager.h>
+
+uint32_t wait_pll_lock(void)
+{
+	uint32_t data;
+	uint32_t count = 0;
+
+	do {
+		data = mmio_read_32(CLKMGR_OFFSET + CLKMGR_STAT);
+		count++;
+		if (count >= 1000)
+			return -ETIMEDOUT;
+
+	} while ((CLKMGR_STAT_MAINPLLLOCKED(data) == 0) ||
+			(CLKMGR_STAT_PERPLLLOCKED(data) == 0));
+	return 0;
+}
+
+uint32_t wait_fsm(void)
+{
+	uint32_t data;
+	uint32_t count = 0;
+
+	do {
+		data = mmio_read_32(CLKMGR_OFFSET + CLKMGR_STAT);
+		count++;
+		if (count >= 1000)
+			return -ETIMEDOUT;
+
+	} while (CLKMGR_STAT_BUSY(data) == CLKMGR_STAT_BUSY_E_BUSY);
+
+	return 0;
+}
+
+uint32_t pll_source_sync_config(uint32_t pll_mem_offset, uint32_t data)
+{
+	uint32_t val = 0;
+	uint32_t count = 0;
+	uint32_t req_status = 0;
+
+	val = (CLKMGR_MEM_WR | CLKMGR_MEM_REQ |
+		(data << CLKMGR_MEM_WDAT_OFFSET) | CLKMGR_MEM_ADDR);
+	mmio_write_32(pll_mem_offset, val);
+
+	do {
+		req_status = mmio_read_32(pll_mem_offset);
+		count++;
+	} while ((req_status & CLKMGR_MEM_REQ) && (count < 10));
+
+	if (count >= 100)
+		return -ETIMEDOUT;
+
+	return 0;
+}
+
+uint32_t pll_source_sync_read(uint32_t pll_mem_offset)
+{
+	uint32_t val = 0;
+	uint32_t rdata = 0;
+	uint32_t count = 0;
+	uint32_t req_status = 0;
+
+	val = (CLKMGR_MEM_REQ | CLKMGR_MEM_ADDR);
+	mmio_write_32(pll_mem_offset, val);
+
+	do {
+		req_status = mmio_read_32(pll_mem_offset);
+		count++;
+	} while ((req_status & CLKMGR_MEM_REQ) && (count < 10));
+
+	if (count >= 100)
+		return -ETIMEDOUT;
+
+	rdata = mmio_read_32(pll_mem_offset + 0x4);
+
+	return rdata;
+}
+
+void config_clkmgr_handoff(struct handoff *hoff_ptr)
```

Comment:
What is using this function?

Reply:
Hi Carlo,
We plan to support and upstream Agilex features by phases.

There will be 2 kinds of Zephyr boot for Agilex:

(1) ATF BL2 -> ATF BL31 -> Zephyr
     - This is what we upstream now, first batch of of Agilex upstream code

(2) Zephyr self-boot, processor directly boot from zephyr.bin
     - We will upstream this flow later. 
     - We will need to call 'z_arm64_el3_plat_init' function to initialize hardware component (example: pinmux, clock)
     - config_clkmgr_handoff will be used later once we enable self-boot flow. And, the clock_agilex_ll.c is shared/migrated from our ATF source code. Thus, we retains the function for future use.

Kindly advise this is acceptable for the mainline code.  Thanks a lot.

Reply:
If you are upstreaming now only the ATF boot path I would expect only to find in this PR code related to that path. So, please remove any unused / not needed code for this PR, you can add that later when upstreaming the other boot flow.

Reply:
Noted, I will remove all unused code for this phase.